File: src/api/tracker.ts
const TARGET_USER_ID = 'NpFFvozZSFWnCKdmutkISEGPf8o2';
const API_ENDPOINTS = [
  `/api/tracker?userId=${TARGET_USER_ID}`,
  `https://medx.srihari.quest/api/tracker?userId=${TARGET_USER_ID}`,
];
const LOCAL_STORAGE_KEY = 'fmge_tracker_data_v1';

export const SUBJECTS_LIST = [
  'Anatomy', 'Physiology', 'Biochemistry', 'Pathology',
  'Microbiology', 'Pharmacology', 'Forensic medicine',
  'Community Medicine (PSM)', 'General Medicine', 'General Surgery',
  'Obstetrics & Gynecology (OBG)', 'Pediatrics', 'Ophthalmology',
  'Otorhinolaryngology (ENT)', 'Orthopedics', 'Anesthesiology',
  'Dermatology & Venereology', 'Psychiatry', 'Radiodiagnosis (Radiology)',
];

export const SUBJECT_FIELDS = ['Videos', 'R1', 'R2', 'PYQs', 'RevisionVideos', 'Qbank'];
export const TOTAL_ITEMS = 19 * 6 + 7;

export interface TrackerSubject {
  Videos: boolean;
  R1: boolean;
  R2: boolean;
  PYQs: boolean;
  RevisionVideos: boolean;
  Qbank: boolean;
}

export interface TrackerData {
  subjects: Record<string, TrackerSubject>;
  gts: Record<string, boolean>;
}

const INITIAL_STATE: TrackerData = {
  subjects: {},
  gts: { GT1: false, GT2: false, GT3: false, GT4: false, GT5: false, GT6: false, GT7: false },
};

SUBJECTS_LIST.forEach(sub => {
  INITIAL_STATE.subjects[sub] = {
    Videos: false, R1: false, R2: false, PYQs: false, RevisionVideos: false, Qbank: false,
  };
});

function getLocalData(): TrackerData | null {
  try {
    const saved = localStorage.getItem(LOCAL_STORAGE_KEY);
    if (saved) {
      const parsed = JSON.parse(saved);
      return {
        subjects: { ...INITIAL_STATE.subjects, ...(parsed.subjects || {}) },
        gts: { ...INITIAL_STATE.gts, ...(parsed.gts || {}) },
      };
    }
  } catch (e) {
    console.error('Error reading tracker local storage:', e);
  }
  return null;
}

function saveLocalData(data: TrackerData) {
  try {
    localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(data));
  } catch (e) {
    console.error('Error saving tracker local storage:', e);
  }
}

export function subscribeTrackerData(callback: (data: TrackerData) => void): () => void {
  let isCancelled = false;

  const fetchLatest = async () => {
    if (isCancelled) return;
    const data = await getTrackerData();
    if (!isCancelled) {
      callback(data);
    }
  };

  fetchLatest();
  const intervalId = setInterval(fetchLatest, 5000);

  return () => {
    isCancelled = true;
    clearInterval(intervalId);
  };
}

export async function getTrackerData(): Promise<TrackerData> {
  for (const url of API_ENDPOINTS) {
    try {
      const res = await fetch(url);
      if (res.ok) {
        const data = await res.json();
        const merged: TrackerData = {
          subjects: { ...INITIAL_STATE.subjects, ...(data.subjects || {}) },
          gts: { ...INITIAL_STATE.gts, ...(data.gts || {}) },
        };
        saveLocalData(merged);
        return merged;
      }
    } catch (error) {
      // try next endpoint
    }
  }
  return getLocalData() || INITIAL_STATE;
}

export async function updateSubjectTracker(subject: string, field: string, value: boolean): Promise<void> {
  const current = getLocalData() || INITIAL_STATE;
  const updatedSubjects = {
    ...current.subjects,
    [subject]: {
      ...(current.subjects[subject] || { Videos: false, R1: false, R2: false, PYQs: false, RevisionVideos: false, Qbank: false }),
      [field]: value,
    },
  };
  const updatedData: TrackerData = { ...current, subjects: updatedSubjects };
  saveLocalData(updatedData);

  for (const url of API_ENDPOINTS) {
    try {
      const baseUrl = url.split('?')[0];
      await fetch(baseUrl, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ userId: TARGET_USER_ID, subject, field, value }),
      });
      break;
    } catch (error) {
      // fallback
    }
  }
}

export async function updateGTTracker(gt: string, value: boolean): Promise<void> {
  const current = getLocalData() || INITIAL_STATE;
  const updatedGts = {
    ...current.gts,
    [gt]: value,
  };
  const updatedData: TrackerData = { ...current, gts: updatedGts };
  saveLocalData(updatedData);

  for (const url of API_ENDPOINTS) {
    try {
      const baseUrl = url.split('?')[0];
      await fetch(baseUrl, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ userId: TARGET_USER_ID, gt, value }),
      });
      break;
    } catch (error) {
      // fallback
    }
  }
}

export function calculateProgress(data: TrackerData): number {
  let count = 0;
  SUBJECTS_LIST.forEach(sub => {
    const s = data.subjects[sub];
    if (s) {
      SUBJECT_FIELDS.forEach(f => { if ((s as any)[f]) count++; });
    }
  });
  Object.keys(data.gts).forEach(gt => {
    if (data.gts[gt]) count++;
  });
  return Math.round((count / TOTAL_ITEMS) * 100) || 0;
}

File: ios/Shared/Models/Tracker.swift
import Foundation

/// The 19 FMGE subjects, in the order the tracker grid shows them
/// (`SUBJECTS_LIST`, src/api/tracker.ts:8).
enum Syllabus {
    static let subjects = [
        "Anatomy", "Physiology", "Biochemistry", "Pathology",
        "Microbiology", "Pharmacology", "Forensic medicine",
        "Community Medicine (PSM)", "General Medicine", "General Surgery",
        "Obstetrics & Gynecology (OBG)", "Pediatrics", "Ophthalmology",
        "Otorhinolaryngology (ENT)", "Orthopedics", "Anesthesiology",
        "Dermatology & Venereology", "Psychiatry", "Radiodiagnosis (Radiology)",
    ]

    static let grandTests = (1...7).map { "GT\($0)" }

    /// Six checkboxes per subject, plus seven grand tests: 19 × 6 + 7 = 121.
    static let totalItems = subjects.count * SubjectField.allCases.count + grandTests.count

    /// Exam date the countdowns tick towards (StudyPage.tsx:60).
    static let examDate: Date = {
        var comps = DateComponents()
        comps.year = 2027
        comps.month = 1
        comps.day = 9
        return Calendar.current.date(from: comps) ?? Date()
    }()
}

/// One column of the subject grid. Raw values are the API's field names; the
/// labels are the abbreviations the grid shows (`FIELD_LABELS`, StudyPage.tsx:25).
enum SubjectField: String, CaseIterable, Identifiable, Codable, Sendable {
    case videos = "Videos"
    case r1 = "R1"
    case r2 = "R2"
    case pyqs = "PYQs"
    case revisionVideos = "RevisionVideos"
    case qbank = "Qbank"

    var id: String { rawValue }

    var label: String {
        switch self {
        case .videos: return "Videos"
        case .r1: return "Rev 1"
        case .r2: return "Rev 2"
        case .pyqs: return "PYQs"
        case .revisionVideos: return "Rev Vids"
        case .qbank: return "Qbank"
        }
    }
}

struct TrackerSubject: Codable, Hashable, Sendable {
    var videos = false
    var r1 = false
    var r2 = false
    var pyqs = false
    var revisionVideos = false
    var qbank = false

    enum CodingKeys: String, CodingKey {
        case videos = "Videos"
        case r1 = "R1"
        case r2 = "R2"
        case pyqs = "PYQs"
        case revisionVideos = "RevisionVideos"
        case qbank = "Qbank"
    }

    init() {}

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        videos = c.flexBool(.videos)
        r1 = c.flexBool(.r1)
        r2 = c.flexBool(.r2)
        pyqs = c.flexBool(.pyqs)
        revisionVideos = c.flexBool(.revisionVideos)
        qbank = c.flexBool(.qbank)
    }

    subscript(field: SubjectField) -> Bool {
        get {
            switch field {
            case .videos: return videos
            case .r1: return r1
            case .r2: return r2
            case .pyqs: return pyqs
            case .revisionVideos: return revisionVideos
            case .qbank: return qbank
            }
        }
        set {
            switch field {
            case .videos: videos = newValue
            case .r1: r1 = newValue
            case .r2: r2 = newValue
            case .pyqs: pyqs = newValue
            case .revisionVideos: revisionVideos = newValue
            case .qbank: qbank = newValue
            }
        }
    }

    var completedCount: Int {
        SubjectField.allCases.reduce(0) { $0 + (self[$1] ? 1 : 0) }
    }

    var isComplete: Bool { completedCount == SubjectField.allCases.count }
}

struct TrackerData: Codable, Hashable, Sendable {
    var subjects: [String: TrackerSubject] = Syllabus.subjects.reduce(into: [:]) {
        $0[$1] = TrackerSubject()
    }
    var gts: [String: Bool] = Syllabus.grandTests.reduce(into: [:]) { $0[$1] = false }

    enum CodingKeys: String, CodingKey { case subjects, gts }

    init() {}

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)

        var merged: [String: TrackerSubject] = Syllabus.subjects.reduce(into: [:]) {
            $0[$1] = TrackerSubject()
        }
        if let sent = try? c.decodeIfPresent([String: TrackerSubject].self, forKey: .subjects) {
            for (key, value) in sent { merged[key] = value }
        }
        subjects = merged

        var mergedGts: [String: Bool] = Syllabus.grandTests.reduce(into: [:]) { $0[$1] = false }
        if let sent = try? c.decodeIfPresent([String: Bool].self, forKey: .gts) {
            for (key, value) in sent { mergedGts[key] = value }
        }
        gts = mergedGts
    }

    func subject(_ name: String) -> TrackerSubject {
        subjects[name] ?? TrackerSubject()
    }

    // ── Derived ─────────────────────────────────────────────────────────────

    /// Checked items out of 121. Ported from `calculateProgress`
    /// (src/api/tracker.ts:160).
    var completedItems: Int {
        let subjectCount = Syllabus.subjects.reduce(0) { $0 + subject($1).completedCount }
        let gtCount = gts.values.filter { $0 }.count
        return subjectCount + gtCount
    }

    /// Whole percent, matching the web app's `Math.round`.
    var progressPercent: Int {
        guard Syllabus.totalItems > 0 else { return 0 }
        return Int((Double(completedItems) / Double(Syllabus.totalItems) * 100).rounded())
    }

    var completedSubjects: Int {
        Syllabus.subjects.filter { subject($0).isComplete }.count
    }

    var completedGTs: Int { gts.values.filter { $0 }.count }

    /// Home-widget status word (FmgeProgressWidget.tsx:69).
    var statusLabel: String {
        switch progressPercent {
        case ..<10: return "Just starting"
        case ..<30: return "Getting going"
        case ..<60: return "Halfway there"
        case ..<90: return "Almost done"
        default: return "Ready!"
        }
    }

    var daysToExam: Int {
        max(0, DayKey.calendarDays(from: Date(), to: Syllabus.examDate))
    }
}
